### ISSUE TITLE
Mechanism to refresh token

### ISSUE THREAD

**Issue by baldred** [godaddy/kubernetes-client#328]
I'm wondering if there is any way we can use this kubernetes-client module to refresh authentication token when **auth-provider** is **oidc**. Below is the snippet of my kubeconfig (it's testing environment so it's okay to share in plaintext).

```
auth-provider:
      name: oidc
      config:
        client-id: bx
        client-secret: bx
        id-token: eyJraWQiOiIyMDE3MTAzMC0wMDowMDowMCIsImFsZyI6IlJTMjU2In0.eyJpYW1faWQiOiJJQk1pZC0zMTAwMDIxMlcxIiwiaXNzIjoiaHR0cHM6Ly9pYW0ubmcuYmx1ZW1peC5uZXQva3ViZXJuZXRlcyIsInN1YiI6ImJhbGRyZWRAaWQuaWJtLmNvbSIsImF1ZCI6ImJ4IiwiZ2l2ZW5fbmFtZSI6IkFsZHJlZCIsImZhbWlseV9uYW1lIjoiQmVuZWRpY3QiLCJuYW1lIjoiQWxkcmVkIEJlbmVkaWN0IiwiZW1haWwiOiJiYWxkcmVkQGlkLmlibS5jb20iLCJleHAiOjE1MzY1NzExMzEsInNjb3BlIjoiaWJtIG9wZW5pZCIsImlhdCI6MTUzNjU2NzUzMX0.S08p2PqmtLezWkPDJX7K-P54QMRmVKooMAKkL8gf4PL5NkLBV-zhumt3xmHzU2nN34Jkw9hlnALiEdDWY5ZC5a-ObXDy-nJpE65InvdF-icXQZknVOjqiKrrTDI5QR_a67vSrO1dgmDmjQ0Al7Nw9YxIqZPU5gKq5CSZcqqW_X9QYQl5sp5Qaec3wgY4ALOnnzuqQ38TwmHC_4P3QU0HF10Tz7f2nwYxSMyxwX8f5WGcanymZ8Mf7oSiCUMPEjUY9rnxTxEzQA-2Sf_Dv80dcQWZCXXwE4tgVbkUls98B3xDVaCsgs6WhplC87P1BaSrEMmHrn2nDfZKUwANcCwaWg
        idp-issuer-url: https://iam.ng.bluemix.net/kubernetes
        refresh-token: J1CGY45D33Ic166oaf1IFxA9_D9JtwBf7IUYssi73imk8M4DE81vHh0dPhQ3BlxFc4qhovygznn0N0hpuK-AUNbBS-Xd3o4D4qrxJ6Ugxe3PE2DyN5S1zdeFPHrk4eVSYEIT1tDH7II6MgzavqjfkvcvobG0j0Mj8Lzpw5C_Wdl6-4zJiS-qFVsAL42IG3ASJCcx180pwgHulFYD8NlvjtbttbDDD6Yzq70Y0N1LML7VDbaFPsVD3m4poNSu4Csdd9p6fFCearOioGyUub28RNuH26PItJgQXpN_pQnSOujVpAh6yeI6E-Ivb7NOps1t_xZLbwuij3ypf6qZIZqX1OFlmGnNiAxbIXCE5NuUhAbEoeWvGp_J_K4AU_GqiVHmuya94y_E8leEy_sreVtU-U4k5RfovIevn6nBj7_b7Dc55-Q-pxjFov2LRGF1Xmhr3m4RVMmtSnzvVYcvTLDDSfjd2aXpEBjhd2Ht8f7HMF4K63_QIG_rpXuWI_OsyXPzd9x-I8vr-kRt_qWpTEG71OcOlONwKbsSZgSQtpPCllL4lvySBL4UvHz4re7tntOX0buCLnFwDy_EQthvyR6LIk37XTlv_Bp4OJPN2DNYgr0P6CRtGNCHmH7lwksfF_AeAbj4wV2albLHN36H3QKa0dpGV3q_Ob8BWT9jg83-H11-AEyLIDBp-QcbJyHgUI4LN93yp3aHl1TxPfPBI8nU46fp0q8MU7ftbgXmE1M3z29ZlN119goEYhlvRoFaB6XujUz9hrdjTws0RuFMGjciKn3JWdhcQUaT-tt8ys4JqC4IcLP7E44tI1qqHuy7bZ3pMOYIzi5ovytcK3qVOZP0BDINGDNVzi31KyCyRLbGYBVRHBSZipnfdKW06fCr8LdNc4U
```

I took at a look at the code (https://github.com/godaddy/kubernetes-client/blob/master/lib/request.js). It seems to me that there is a mechanism for refresh. But if I load kubeconfig with expired token, the request will just time-out:

```
(node:12517) UnhandledPromiseRejectionWarning: AggregateError: 
    RequestError: Request timed out
at timeoutFn (/Users/aldred/Workspace/blockchain/kubernetes-api/node_modules/got/index.js:333:53)
        at asPromise (/Users/aldred/Workspace/blockchain/kubernetes-api/node_modules/got/index.js:411:18)
```

**Comment by silasbw** [comment]
hey @baldred, thanks for tracking this down.  So, the problem you're seeing is that the default timeout value is too small for your needs, and your refresh request is timing out?  We should def fix that -- I'm wondering if it's by hardcoding a value, like your patch does; or by providing a way to configure it.

One approach would be to extend https://github.com/godaddy/kubernetes-client/blob/master/lib/request.js#L40 with a timeout option and pass that value through the auth backend.  You need to pass the timeout value when constructing your kubernetes-client instance.  Would something like that work for you?

**Comment by baldred** [comment]
Hi @silasbw correct, the default timeout is too small for my needs. I think passing the timeout value when constructing the instance will be a great idea. 

The other thing is that I found out even if the token is refreshed, the next request will still use the outdated token and a refresh will be triggered. I would writing the new token back into the request instance, by adding `this.requestOptions.auth = newAuth;` in `request.js`:

```
request(method, options, cb) {
    const uri = typeof options.path === 'string' ? options.path : options.path.join('/');
    const requestOptions = Object.assign({
      method: method,
      uri: uri,
      body: options.body,
      json: true,
      qs: options.qs,
      headers: options.headers
    }, this.requestOptions);

    if (typeof cb !== 'function') return request(requestOptions);

    const auth = this.authProvider;
    return request(requestOptions, (err, res, body) => {
      if (err) return cb(err);

      // Refresh auth if 401
      if (res.statusCode === 401 && auth.type) {
        return refreshAuth(auth.type, auth.config)
          .then(newAuth => {
            requestOptions.auth = newAuth;
            this.requestOptions.auth = newAuth; // add this
            return request(requestOptions, (err, res, body) => {
              if (err) return cb(err);
              return cb(null, { statusCode: res.statusCode, body });
            });
          })
          .catch(err => cb(err));
      }

      return cb(null, { statusCode: res.statusCode, body: body });
    });
  }
```

**Comment by jonathan-yk-tan** [comment]
@silasbw is this update in progress?

**Comment by silasbw** [comment]
I'm sorry for the latency of this fix :pray: . How does this look?

https://github.com/godaddy/kubernetes-client/pull/339

I need docs, but I added an example to illustrate usage.

**Comment by baldred** [comment]
hi @silasbw should be okay :)..please merge to master..thanks a lot for your help.

**Comment by silasbw** [comment]
... I also published kubernetes-client@6.3.0 to npmjs.

Let me know how that works for you!
